From 0423a5079be3a146c424993df0d4bd997b81b4f0 Mon Sep 17 00:00:00 2001
From: Wang Wensheng <wangwensheng4@huawei.com>
Date: Wed, 23 Aug 2023 15:12:25 +0800
Subject: [PATCH 50/53] [Huawei] mm/sharepool: Use sp_area_put_locked() if
 possible

Offering: HULK
hulk inclusion
category: feature
bugzilla: N/A

--------------------------------

This we avoid taking the spg->rw_lock immediately after releasing it.

Signed-off-by: Wang Wensheng <wangwensheng4@huawei.com>
---
 mm/share_pool.c | 34 +++++++++++++---------------------
 1 file changed, 13 insertions(+), 21 deletions(-)

diff --git a/mm/share_pool.c b/mm/share_pool.c
index defb5bee6445..ff3baf415842 100644
--- a/mm/share_pool.c
+++ b/mm/share_pool.c
@@ -1733,17 +1733,6 @@ static void sp_area_put_locked(struct sp_area *spa)
 	}
 }
 
-static void sp_area_put(struct sp_area *spa)
-{
-	if (atomic_dec_and_test(&spa->use_count)) {
-		down_write(&spa->spg->rw_lock);
-		if (!RB_EMPTY_NODE(&spa->spg_link))
-			sp_group_delete_area(spa->spg, spa);
-		up_write(&spa->spg->rw_lock);
-		sp_area_free(spa);
-	}
-}
-
 void sp_area_drop(struct vm_area_struct *vma)
 {
 	struct sp_area *spa = vma->vm_private_data;
@@ -1760,7 +1749,13 @@ void sp_area_drop(struct vm_area_struct *vma)
 	if (!spa)
 		return;
 
-	sp_area_put(spa);
+	if (atomic_dec_and_test(&spa->use_count)) {
+		down_write(&spa->spg->rw_lock);
+		if (!RB_EMPTY_NODE(&spa->spg_link))
+			sp_group_delete_area(spa->spg, spa);
+		up_write(&spa->spg->rw_lock);
+		sp_area_free(spa);
+	}
 }
 
 /*
@@ -1855,22 +1850,19 @@ static int sp_free_inner(unsigned long addr, int spg_id, bool is_sp_free)
 	down_write(&spg->rw_lock);
 	spa = sp_area_get(spg, addr);
 	if (!spa) {
-		up_write(&spg->rw_lock);
 		pr_debug("%s, invalid input addr %lx\n", str, addr);
 		ret = -EINVAL;
-		goto out;
+		goto drop_spg;
 	}
 
 	if ((is_sp_free && spa->type != SPA_TYPE_ALLOC) ||
 	    (!is_sp_free && spa->type == SPA_TYPE_ALLOC)) {
-		up_write(&spg->rw_lock);
 		ret = -EINVAL;
 		pr_warn("%s failed, spa_type is not correct\n", str);
 		goto drop_spa;
 	}
 
 	if (!current->mm && spa->applier != current->tgid) {
-		up_write(&spg->rw_lock);
 		ret = -EPERM;
 		pr_err("%s, free a spa allocated by other process(%d), current(%d)\n",
 			str, spa->applier, current->tgid);
@@ -1887,11 +1879,11 @@ static int sp_free_inner(unsigned long addr, int spg_id, bool is_sp_free)
 
 	if (current->mm)
 		update_mem_usage(spa_size(spa), false, spa->is_hugepage, spg_node, spa->type);
-	up_write(&spg->rw_lock);
 
 drop_spa:
-	sp_area_put(spa);
-out:
+	sp_area_put_locked(spa);
+drop_spg:
+	up_write(&spg->rw_lock);
 	sp_group_put(spg);
 	return ret;
 }
@@ -2228,8 +2220,8 @@ void *__mg_sp_alloc_nodemask(unsigned long size, unsigned long sp_flags, int spg
 	if (!ret)
 		update_mem_usage(spa_size(spa), true, spa->is_hugepage, spg_node, spa->type);
 
+	sp_area_put_locked(spa);
 	up_write(&spg->rw_lock);
-	sp_area_put(spa);
 
 out:
 	sp_group_put(spg);
@@ -2483,8 +2475,8 @@ void *mg_sp_make_share_k2u(unsigned long kva, unsigned long size,
 		update_mem_usage(spa_size(spa), true, spa->is_hugepage, ori_node, spa->type);
 		vmalloc_area_set_flag(kc.kva_aligned, VM_SHAREPOOL);
 	}
+	sp_area_put_locked(spa);
 	up_write(&spg->rw_lock);
-	sp_area_put(spa);
 	sp_group_put(spg);
 
 	return ret ? ERR_PTR(ret) : (void *)(spa->va_start + (kc.kva - kc.kva_aligned));
-- 
2.17.1

